entities.py

```diff
@@ -899,13 +726,10 @@ def die(self):
         Dies.
         """
         print(f"{self.name} has died.")
-        if relics['Gremlin Horn'] in player.relics:
+        if items.relics['Gremlin Horn'] in player.relics:
             player.energy += 1
             player.draw_cards(True, 1)
-        try:
-            active_enemies.remove(self)
-        except ValueError:
-            raise Exception(f"{self.name} is not in the active enemies list.")
+        self.state = EnemyState.DEAD
```

Comment:
@miketwo Any idea how to get death messages from here to the combat instance?

Reply:
You can publish a death message if you want other things in general to react to that (like relics and stuff). 

`Combat` knows what enemies are fighting -- why can't it just check their state?

Reply:
That sounds good. Here's other options, just to be exhaustive:

You could technically have combat register for messages too, but I find that it's better to avoid backflow in pub-sub architectures. When you publish, you're trying _NOT_ to have a specific receiver in mind. (Otherwise, it would make more sense to just call the receiver directly.) 

A callback is another option. Combat initializes a bunch of characters to the fight (by adding them to active_enemies and whatnot). You could pass down a function to each enemy that gets called on death. This is a "callback" -- it's typically used in async code (code that uses multiple threads or processors), but if you're careful it can be used here.

So when initializing an enemy, it'd be something like:
```python
SomeEnemy(..., on_death_callback=self.on_death)
```

You can default that callback to `None`. And then when an enemy is dying, if the callback is not `None`, it can call it, maybe with itself as the parameter. 

This allows Combat to instantly do something when an enemy dies, but you'd have to be careful about what you do, because it's not async code. So the call stack is going to return to wherever it was -- at best you could increment a counter or change a flag or something that gets checked in the combat loop.

Reply:
That sounds good. Here's other options, just to be exhaustive:

You could technically have combat register for messages too, but I find that it's better to avoid backflow in pub-sub architectures. When you publish, you're trying _NOT_ to have a specific receiver in mind. (Otherwise, it would make more sense to just call the receiver directly.) 

A callback is another option. Combat initializes a bunch of characters to the fight (by adding them to active_enemies and whatnot). You could pass down a function to each enemy that gets called on death. This is a "callback" -- it's typically used in async code (code that uses multiple threads or processors), but if you're careful it can be used here.

So when initializing an enemy, it'd be something like:
```python
SomeEnemy(..., on_death_callback=self.on_death)
```

You can default that callback to `None`. And then when an enemy is dying, if the callback is not `None`, it can call it, maybe with itself as the parameter. 

This allows Combat to instantly do something when an enemy dies, but you'd have to be careful about what you do, because it's not async code. So the call stack is going to return to wherever it was -- at best you could increment a counter or change a flag or something that gets checked in the combat loop.